[3.1] feature: allow specific object create templates in embedded object create modal (#8716)
nautobot/nautobot: nautobot/docs/development/apps/migration/embedded_actions.md

# Migration Guide: Embedded Actions

## Background: the shift to Embedded Actions

In 3.1 release, Nautobot has significantly improved the user experience of creating and updating objects by introducing embedded create and search actions. Historically, if a user was filling out a form and needed to select a related object that didn't exist yet (e.g., needing to create a new Manufacturer while adding a Device Type), they had to navigate away, create the object, and then return — often losing their progress.

To solve this, Nautobot 3.1 introduced:

- Embedded Search: Allowing users to perform advanced filtering and selection of related objects dynamically within a modal.
- Embedded Create: Permitting users to create new related objects on the fly via a modal without ever leaving their current page or losing their form context.

> This shift toward a more dynamic, single-page-like experience fundamentally changes how forms are rendered and injected into the DOM.

Reviewer: ```suggestion
```

!!! info
    Embedded Actions were introduced only in standard object add/create/update/edit form views which inherit from generic Nautobot form and view or vieweset classes. Any custom form views or other non-CRUD forms (such as Jobs) were not affected in any way by this change.

## Overview of the new API

Nautobot 3.1 introduces support for rendering specific object creation templates directly within the embedded object creation modals.

Because these embedded forms are loaded dynamically into the DOM via AJAX after the initial page load, any JavaScript that traditionally relied on `$(document).ready()` or `DOMContentLoaded` event will not execute for the newly loaded modal content. Moreover, unscoped JavaScript code executed on page load may potentially leak into and unwillingly linger in the global `window` object, causing bugs. To resolve this and give developers a clean way to hook into the form lifecycle, we have introduced a new frontend API: `window.nb.form`.

This guide explains how Nautobot App developers should update their form-related JavaScript to remain fully compatible with both standard page loads and dynamic modals.

## Why the change?

> 

Reviewer: ```suggestion
```

Previously, when you wrote JavaScript to enhance an app's form (e.g., toggling field visibility based on user input, or setting up dynamic formsets), you likely wrapped your logic in a global document-ready block.

When a user opens a dynamic modal to create an object, the modal content is fetched and injected into the page. Because the document is already "ready", those initialization events do not re-fire for the dynamically injected content, leaving your custom form logic uninitialized and broken.

The new `window.nb.form` API provides a standardized lifecycle event that dispatches both on the initial page load and whenever a dynamic form is loaded into a modal.

> 

Reviewer: we should have a section on how to just "turn it off"

Author: good point, added in 8dcb21bfb49118ee999dda96388805e82366b2af

> ##Migrating to `window.nb.form`

Reviewer 2: ```suggestion
## Migrating to `window.nb.form`
```

        const getField = (name) => document.querySelector(window.nb.form.getFieldAutoId('{{ form.auto_id }}', name));
        getField('field_name').doSomething();
        getField('other_field_name').doSomethingElse();
    });
    ```

### 5. Let `jsify_form` handle Select2 initialization for standard fields

Whenever a form is loaded — whether on the initial page load or dynamically inside an embedded action modal — Nautobot automatically calls a core function named `jsify_form`. This function is responsible for initializing standard Nautobot UI components, which natively includes applying Select2 to certain dropdowns.

You should remove manual `.select2()` calls on standard Nautobot fields to prevent conflicts or double-initialization bugs.

!!! note
    If your app introduces specific, highly customized fields that `jsify_form` does not process by default, you may still need to manually call `.select2()` on those specific custom fields inside your `nb-form:load:{{ obj_type }}` event listener.

## Summary checklist for App Developers

- Read up on the transition to embedded search and create forms to understand the new UX paradigm.

> - Identify all JavaScript files in your app that interact with add/create/update/edit forms.

Reviewer 2: ```suggestion
- Identify all JavaScript content in your app that interacts with add/create/update/edit forms.
```

When a user opens a dynamic modal to create an object, the modal content is fetched and injected into the page. Because the document is already "ready", those initialization events do not re-fire for the dynamically injected content, leaving your custom form logic uninitialized and broken.

The new `window.nb.form` API provides a standardized lifecycle event that dispatches both on the initial page load and whenever a dynamic form is loaded into a modal.

## Controlling Embedded Actions on forms

While the embedded search and create modals significantly improve the default user experience, there are times when they might not be appropriate for your app's specific workflow. For instance, you might have a highly customized field where a generic creation modal doesn't capture the necessary context, or a field where you explicitly want to prevent users from creating new related objects on the fly.

To give developers granular control over the UI, Nautobot forms and fields support several attributes to explicitly enable or disable these embedded action buttons on a per-field basis. You can define these attributes directly on your form class:

### 1. Form meta class attributes

> You can define white- and blacklists on your form meta class to control which fields get embedded action buttons:

Reviewer 2: nit: prefer `inclusion list` and `exclusion list` terminology over `whitelist` and `blacklist`.

Author: fair, changed in 3222f375047f206c7a855db089ca185602c3ba05
